docs/source/boundary_conditions.rst

```diff
@@ -310,3 +310,190 @@ solved with boundary conditions as follows:
+
+where `\mathrm{F}(u)_i = \hat F(u; \phi_i)` and
+`J` is the Jacobian matrix defined by the Gâteaux derivative of
+`F`:
+
+.. math::
+
+   dF(u; \tilde{u}, v) = \lim_{h\rightarrow0}
+   \frac{\hat F(u+h\tilde u; v) - \hat F(u; v)}{h} \quad \forall v,
+   \tilde u \in V
+
+The actual Jacobian matrix is given by:
+ 
+.. math::
+
+   J_{ij} = dF(u; \phi_i, \phi_j)
+
+where `\phi_i`, `\phi_j` are the ith and jth 
+basis functions of `V`. Our definition of the modified residual
+`\hat F` produces submatrices of distinct structures on the form 
+boundary condition rows of `J` and on the remaining rows of `J`.
+In other words, the rows of `J` corresponding to the boundary
+condition nodes are replaced by Jacobian matrix corresponding to
+`F_\Gamma`.
+The resulting Jacobian matrix is thus non-symmetric in general.
+
+Contrary to the case of Dirichlet boundary conditions, we do not
+know the values of the solution on the boundary condition nodes a priori,
+and the whole system of equations are to be solved monolithically
+using linear/nonlinear solvers.
+
+
+Implementation
+--------------
+
+Variational problems
+~~~~~~~~~~~~~~~~~~~~
+
+Both linear and nonlinear PDEs are solved in residual form in
+Firedrake using the `PETSc SNES interface <http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/SNES/>`_. In the case of linear
+systems, a single step of Newton is employed. 
+
+In the following we will use ``F`` for the residual :class:`~ufl.form.Form`
+and ``J`` for the Jacobian :class:`~ufl.form.Form`. In both cases these
+forms do not include the boundary conditions. 
+A Form boundary condition :class:`~.FormBC` object separately carries ``F_{Form}`` 
+for the boundary residual :class:`~ufl.form.Form` and ``J_{Form}`` for the 
+boundary Jacobian :class:`~ufl.form.Form`.
+Additionally ``u`` will be the solution :class:`~.Function`.
+
+Form boundary conditions are applied as follows:
+
+1. Each time the solver assembles the Jacobian matrix, the following happens. 
+
+   a) ``J`` is assembled using modified indirection maps in which the
+      row indices associated with form boundary condition node have been replaced by negative
+      values. PETSc interprets these negative indices as an
+      instruction to drop the corresponding entry. 
+
+   b) ``J_{Form}`` is assembled to populate the form boundary node rows
+      that are not populated in a).
+   
+2. Each time the solver assembles the residual, the following happens.
+   
+   a) ``F`` is assembled using unmodified indirection maps taking no
+      account of the boundary conditions. This results in an assembled
+      residual which is correct on the non-boundary condition nodes but
+      contains spurious values in the boundary condition entries.
+
+   b) The entries of ``F`` corresponding to boundary condition nodes
+      are set to zero.
+
+   c) ``F_{Form}`` is assembled to populate the entries corresponding to 
+      the form boundary condition nodes that have been zeroed in b).
```

Comment:
OK, so the big change to assembly is that split assembly of operators and residuals that are subject to equational boundary conditions into two pieces. We do this with `DirichletBC` too, except that the subsequent "assembly" that we do to fill in the bits is trivial.

Does this suggest an implementation where the presence of BCs in a variational problem/assembly turns (internally) into a 2-tuple of `(before_action, after_action)`? This would allow us to redo this code to avoid the action-at-a-distance of setting `bc.f` in FormBCs.

Reply:
For now I will make an alternative change to avoid bc.f, but will come back to this idea when I have to consider more general cases where boundary equations themselves require boundary conditions (on edges, say).